torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py

```diff
@@ -0,0 +1,110 @@
+from typing import Any, Callable, List, Optional
+
+import torch
+import torch.distributed as dist
+from torch.distributed.optim import ZeroRedundancyOptimizer
+from torch.distributed.optim.zero_redundancy_optimizer import _OverlapStatus
+from torch.nn.parallel.distributed import DistributedDataParallel
+
+
+def hook_then_zero_step(
+    hook: Callable[[Any, dist.GradBucket], torch.futures.Future],
+    ddp: DistributedDataParallel,
+    zero: ZeroRedundancyOptimizer,
+) -> Callable[[Any, dist.GradBucket], torch.futures.Future]:
+    r"""
+    Modifies the given ``hook`` to additionally perform a partial
+    :class:`ZeroRedundancyOptimizer` :meth:`step` using the gradients in the
+    :class:`DistributedDataParallel` gradient bucket provided by the ``hook``.
+
+    Arguments:
+        hook (Any * dist.GradBucket -> torch.futures.Future): the hook to
+            modify.
+        ddp (DistributedDataParallel): the :class:`DistributedDataParallel`
+            instance to use.
+        zero (ZeroRedundancyOptimizer): the :class:`ZeroRedundancyOptimizer`
+            instance to use.
+
+    Returns:
+        The modified hook.
+    """
+    assert zero._is_functional_optim, "ZeroRedundancyOptimizer must be " \
+        "constructed with a functional optimizer class to use this function"
+
+    def hook_then_zero_fn(
+        state,
+        bucket: dist.GradBucket,
+    ) -> torch.futures.Future:
+        r"""
+        Returns a :class:`Future` that gives a gradient bucket tensor and
+        performs a partial :class:`ZeroRedundancyOptimizer` :meth:`step` using
+        the gradients in that bucket.
+
+        Arguments:
+            state: any state for the hook.
+            bucket (dist.GradBucket): the :class:`DistributedDataParallel`
+                gradient bucket.
+        """
+        fut = hook(state, bucket)
+
+        def zero_step(_) -> List[torch.Tensor]:
+            r"""
+            Performs a partial :class:`ZeroRedundancyOptimizer` :meth:`step`
+            using the gradients in the given :class:`DistributedDataParallel`
+            gradient bucket.
+            """
+            # Proceed as normal until the DDP buckets have been rebuilt
+            if not ddp._has_rebuilt_buckets:
+                return [bucket.get_tensor()]
+
+            bucket_index = bucket.get_index()
+            rank = zero.global_rank
+            overlap_info = zero._overlap_info
+            if overlap_info.status == _OverlapStatus.UNINITIALIZED:
+                overlap_info.status = _OverlapStatus.DDP_HAS_REBUILT_BUCKETS
+
+            bucket_params = bucket.get_model_params_for_bucket()
+            assert len(bucket_params) > 0, "Empty bucket"
+            rank_to_update = zero._ddp_bucket_index_to_rank(bucket_index)
+
+            # Once DDP buckets have been rebuilt but ZeRO has not been
+            # properly initialized yet, collect the information needed
+            if overlap_info.status == _OverlapStatus.DDP_HAS_REBUILT_BUCKETS:
```

Comment:
I don't see a state transition to initialized, where does that happen?

Reply:
It happens in [`ZeroRedundancyOptimizer.step()`](https://github.com/pytorch/pytorch/blob/f60106a46aecca09ffa75cb13a179833e6aaff2d/torch/distributed/optim/zero_redundancy_optimizer.py#L831). My understanding: Buckets are rebuilt during the second forward pass; the communication hooks collect the necessary information in the second backward pass; and the state transition happens in the second optimizer step.